pulsar-broker/src/main/java/org/apache/pulsar/broker/transaction/pendingack/impl/MLPendingAckStore.java

```diff
@@ -332,14 +408,41 @@ public void run() {
                 while (lastConfirmedEntry.compareTo(currentLoadPosition) > 0 && fillEntryQueueCallback.fillQueue()) {
                     Entry entry = entryQueue.poll();
                     if (entry != null) {
-                        ByteBuf buffer = entry.getDataBuffer();
                         currentLoadPosition = PositionImpl.get(entry.getLedgerId(), entry.getEntryId());
-                        PendingAckMetadataEntry pendingAckMetadataEntry = new PendingAckMetadataEntry();
-                        pendingAckMetadataEntry.parseFrom(buffer, buffer.readableBytes());
-                        currentIndexLag.incrementAndGet();
-                        handleMetadataEntry(new PositionImpl(entry.getLedgerId(), entry.getEntryId()),
-                                pendingAckMetadataEntry);
-                        pendingAckReplyCallBack.handleMetadataEntry(pendingAckMetadataEntry);
+                        List<PendingAckMetadataEntry> logs = deserializeEntry(entry);
+                        if (logs.isEmpty()){
+                            continue;
+                        } else if (logs.size() == 1){
+                            currentIndexLag.incrementAndGet();
+                            PendingAckMetadataEntry log = logs.get(0);
+                            handleMetadataEntry(new PositionImpl(entry.getLedgerId(), entry.getEntryId()), log);
+                            pendingAckReplyCallBack.handleMetadataEntry(log);
+                        } else {
+                            /**
+                             * 1. Query batch index of current entry from cursor.
+                             * 2. Filter the data which has already ack.
+                             * 3. Build batched position and handle valid data.
+                             */
+                            long[] ackSetAlreadyAck = cursor.getDeletedBatchIndexesAsLongArray(
+                                    PositionImpl.get(entry.getLedgerId(), entry.getEntryId()));
+                            BitSetRecyclable bitSetAlreadyAck = null;
+                            if (ackSetAlreadyAck != null){
+                                bitSetAlreadyAck = BitSetRecyclable.valueOf(ackSetAlreadyAck);
+                            }
```

Comment:
pending ack clear log data is use markDelete, so we don't need this logic

Reply:
Already fixed.